### PR TITLE
Game smart contract

### PR DESCRIPTION
**Pull request by slavingia**
- [x] Players can seek verification
- [x] Judges can verify winners
- [x] Owner can add and remove judges
- [x] Players can win upon verification, and receive 1,000 FWEB3 tokens
- [x] Tests
- [x] Trophy smart contract that can use the winnerOnly function of the game (once deployed)

### CODE REVIEW

**Comment by mfts** [comment]
I'm currently trying to implement 2 and 3 from the list above but I keep running into the same issue: **smart contracts don't have access to transaction history**, only current state. 

Solution would be to get the transaction data off-chain and then use them on-chain.

**Comment by slavingia** [comment]
No off-chain data for this one, would defeat the point. Should be on-chain. We can rewrite/add new contracts to make it possible.

File: contracts/game.sol
Comment (Bschuster3434) [inline contracts/game.sol:92]: I would encourage you to change from using Arrays to using mappings for this. The problem with these functions is that if the number of people looking to verify at the same time (which is likely), this cost to run this transaction will be super high. I’m talking about orders of magnitude more expensive. Review how the original mapping works and see if you can use that instead.

File: contracts/game.sol
Comment (Bschuster3434) [inline contracts/game.sol:122]: Same comment here. With a long enough array, this will become very expensive quickly. Mappings would be a better way to do this to keep the cost down and keep the same structure.

File: contracts/game.sol
Comment (Bschuster3434) [inline contracts/game.sol:87]: Not all winners can be judges. Are you okay to be the only one who can add judges to this? Not a bad idea, just make sure you don’t bottleneck this too much so the community can continue to verify folks.

File: contracts/game.sol
Comment (Bschuster3434) [inline contracts/game.sol]: Why do we want to stop people who have too many tokens from winning? Maybe instead of making this a requirements, you can still be a winner but you just won’t get a token distribution (I can always just send my tokens to another account if I want to cheat). So either move the token distribution to another function or make it so that this send is nested in an if statement.

File: contracts/game.sol
Comment (Bschuster3434) [inline contracts/game.sol:83]: Need to add some logic around this so not anyone can just reject a winner (should be judgeOnly).

But more importantly, what’s the intention here? If one judge decides not to approve someone, they can reject them from the competition and they have to start over?

File: contracts/game.sol
Comment (Bschuster3434) [inline contracts/game.sol]: So is the intention that only one judge has to allow for approval? Am I reading right? If so, I think we need to have a logic where we have multiple people approving.

**Comment by Bschuster3434** [comment]
Looks like this was done before the rest of the changes. I think it would be worth writing the design of the game at the top (number of approvers, purpose of tokens need, etc.) to make sure I’m firing at the right target.

Biggest issue is really the use of arrays. As the number of winners grows, the potential for the cost to go up increases by quite a lot. Mappings help solve this issue, review how it’s implemented in the other PR.

**Review by Bschuster3434 — CHANGES_REQUESTED** [review]

File: contracts/game.sol
Comment (slavingia) [reply contracts/game.sol:92]: sure, can do!

File: contracts/game.sol
Comment (slavingia) [reply contracts/game.sol]: The reason is for people who've already won by getting tokens. But you're right, people should still be able to "win". Will rethink this.

File: contracts/game.sol
Comment (slavingia) [reply contracts/game.sol]: Yep, currently just one judge is required per approval. I can remove them if there are issues, I don't think we need the complexity of multiple approvals per player.

File: contracts/game.sol
Comment (slavingia) [reply contracts/game.sol:83]: for now they're just removed from the list, so that "spam" doesn't clog up a dashboard we build

File: contracts/game.sol
Comment (Bschuster3434) [inline contracts/game.sol:69]: This may not be the exact error, but this is something that needs to be fixed. The smart contract is sending out ethereum here, but we’re intending to send FWEB3 tokens. Let’s update that.

So, here’s what needs to change:

1. The contract here needs to be gifted an amount of ERC20 tokens (enough to send out). You can do this by giving it an allowance using the allowance function, so that the smart contract has tokens to play with: https://ethereum.org/en/developers/tutorials/understand-the-erc-20-token-smart-contract/
2. This line needs to be updated to do something like _token.transferFrom(address(this), msg.sender, 1000 * 10**18) so that sends out the FWEB3 token.
Might clear the error, let’s see…

File: test/game.js
Comment (Bschuster3434) [inline test/game.js:27]: Add a line to give the game.address an allowance so it have 10000 tokens to play with or something.

File: contracts/game.sol
Comment (Bschuster3434) [inline contracts/game.sol:80]: I’m still not sure how this works exactly. So a single judge can decide to deny someone from making it to the finals? How would that work?

**Review by Bschuster3434 — APPROVED** [review]
Overall, I’d like to see it on a testnet before we got this working in production. But other than that I can’t think of anymore changes I’d make here.

File: contracts/game.sol
Comment (slavingia) [reply contracts/game.sol:80]: sure, i think we can remove this. if we need to hide, we can do it as part of the web2 layer instead

**Comment by slavingia** [comment]
Great. Will start on that tmmrw. Merging so any changes we make will be easier to review

**Pull request merged**
